### PR TITLE
Refactor webpack-dev-server middleware application logic

### PR DESCRIPTION
**Pull request by Swapnilden**
This commit refactors the middleware application logic in the webpack-dev-server codebase to ensure that middleware is applied only once, regardless of how many times it's called. 

Previously, certain middleware, such as static serving middleware, was being applied multiple times to support features like `historyApiFallback`. This refactor eliminates duplicate middleware application by introducing a mechanism to track applied middleware and applying each middleware only once.

Changes:
- Introduced `appliedMiddleware` array to track applied middleware.
- Created `applyMiddlewareOnce` function to apply middleware only if it hasn't been applied before.
- Updated webpack-dev-server codebase to utilize `applyMiddlewareOnce` for applying middleware associated with various features.

This update aims to improve code efficiency and prevent unintended behavior or performance issues caused by duplicate middleware application.

Fixes: #2716

- [ ] This is a **bugfix**
- [ ] This is a **feature**
- [x] This is a **code refactor**
- [ ] This is a **test update**
- [ ] This is a **docs update**
- [ ] This is a **metadata update**

### CODE REVIEW

**Review by alexander-akait — CHANGES_REQUESTED** [review]
Please add test cases

File: lib/Server.js [hunk 1]
```diff
@@ -552,6 +552,35 @@ class Server {
...
+            appliedMiddleware.push(middleware);
+          }
+        }
+        
+        // Apply middleware for each feature only once
+        if (this.options.proxy) {
+          applyMiddlewareOnce.call(this, proxyMiddleware);
+        }
+        
+        if (this.options.contentBase !== false) {
```
Comment (alexander-akait) [inline lib/Server.js:575]: We don't have `contentBase` anymore

**Review by Swapnilden — COMMENTED** [review]
I have included test cases in new commit to ensure that the applyMiddlewareOnce function behaves as expected. Additionally, I have removed references to contentBase, as requested.

File: lib/Server.js [hunk 2]
```diff
@@ -552,6 +552,58 @@ class Server {
...
+            const middlewareFunction1 = jest.fn();
+            const middlewareFunction2 = jest.fn();
+            applyMiddlewareOnce(app, middlewareFunction1);
+            applyMiddlewareOnce(app, middlewareFunction2);
+        
+            expect(app.use).toHaveBeenCalledTimes(2);
+            expect(app.use).toHaveBeenCalledWith(middlewareFunction1);
+            expect(app.use).toHaveBeenCalledWith(middlewareFunction2);
+          });
+        });
```
Comment (alexander-akait) [inline lib/Server.js:604]: We should not store it here

Comment (Swapnilden) [reply lib/Server.js:604]: So please suggest for proceedings

Comment (alexander-akait) [reply lib/Server.js:604]: We have the `test` directory

Comment (Swapnilden) [reply lib/Server.js:604]: can you suggest me file name , it can be big help

**Comment by alexander-akait** [comment]
Closing due to inactivity. Thanks!

**Pull request closed without merging**